## Çoklu Lineer Regresyon (Model)

In [1]:
import pandas as pd
ad = pd.read_csv("advertising.csv", usecols = [1,2,3,4])
df = ad.copy()
df.head()

,TV,radio,newspaper,sales
0,230.1,37.8,69.2,22.1
1,44.5,39.3,45.1,10.4
2,17.2,45.9,69.3,9.3
3,151.5,41.3,58.5,18.5
4,180.8,10.8,58.4,12.9


In [2]:
from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict

In [3]:
X = df.drop("sales", axis = 1)
y = df["sales"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [4]:
X_train.shape

(160, 3)

In [5]:
y_train.shape

(160,)

In [6]:
X_test.shape

(40, 3)

In [7]:
y_test.shape

(40,)

In [8]:
training = df.copy()

In [9]:
training.shape

(200, 4)

#### statsmodels ile model kurulumu

In [10]:
import statsmodels.api as sm
lm = sm.OLS(y_train, X_train)

In [11]:
model = lm.fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              2935.
Date:                Tue, 31 Mar 2026   Prob (F-statistic):                   1.28e-137
Time:                        18:38:57   Log-Likelihood:                         -336.65
No. Observations:                 160   AIC:                                      679.3
Df Residuals:                     157   BIC:                                      688.5
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0531      0.001     36.467      0.000       0.050       0.056
radio          0.2188      0.011     20.138      0.000       0.197       0.240
newspaper      0.0239      0.008      3.011      0.003       0.008       0.040
==============================================================================
Omnibus:                       11.405   Durbin-Watson:                   1.895
Prob(Omnibus):                  0.003   Jarque-Bera (JB):               15.574
Skew:                          -0.432   Prob(JB):                     0.000415
Kurtosis:                       4.261   Cond. No.                         13.5
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Yorum

*R² değeri artmış. Yani satış değişkeninin değişkenliğinin %98'ini açıklayabilmişiz.*

*Yine F istatistiğine ve F-pvalue değerlerine baktığımızda modelin anlamlı olduğunu görüyoruz.*

*Bütün katsayılar anlamlı(p-value değerleri 0'a çok yakın)*

*Yine katsayıları incelediğimizde satış miktarındaki 1 birimlik artışı en çok TV harcamaları, daha sonra newspaper harcamaları, en son radio harcamalarının etkilediğini görüyoruz.*

In [12]:
model.summary().tables[1]

,coef,std err,t,P>|t|,[0.025,0.975]
TV,0.0531,0.001,36.467,0.000,0.050,0.056
radio,0.2188,0.011,20.138,0.000,0.197,0.240
newspaper,0.0239,0.008,3.011,0.003,0.008,0.040


*Tekrar hatırlatmakta fayda var. Model özet çıktısında R² ve F istatistik sonuçları önem arzetmektedir. R² bağımsız değişkenlerin bağımlı değişken üzerindeki etkisi hakkında bilgi verir. F istatistik değeri de modelin anlamlılığını ölçer.*

#### scikit-learn ile model kurulumu

In [13]:
from sklearn.linear_model import LinearRegression 

In [14]:
lm = LinearRegression()
model = lm.fit(X_train, y_train)

In [15]:
#sabit katsayı

In [16]:
model.intercept_

np.float64(2.979067338122629)

In [17]:
#diğer katsayılar

In [18]:
model.coef_

array([0.04472952, 0.18919505, 0.00276111])

## Tahmin

*Model Denklemi:*<br>
*Sales = 2.97 + TV * 0.04 + radio * 0.18 + newspaper * 0.002<br>*

*Örneğin 30 birim TV harcaması, 10 birim radyo harcaması, 40 birim gazete harcaması olduğunda satışların tahmini değeri ne olur ?*

In [19]:
yeni_veri = [[30], [10], [40]]
yeni_veri = pd.DataFrame(yeni_veri).T

In [20]:
model.predict(yeni_veri)

C:\Users\tcmur\AppData\Local\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([6.32334798])

In [21]:
import numpy as np
from sklearn.metrics import mean_squared_error, r2_score

#### Tahmin Başarısının Değerlendirilmesi

In [22]:
#Eğitim hatasının hesaplanması

In [23]:
rmse = np.sqrt(mean_squared_error(y_train, model.predict(X_train)))

In [24]:
rmse

np.float64(1.644727765644337)

*Eğitim setindeki gerçek y değerleri ile eğitim setindeki bağımsız değişkenler üzerinden hesapladığımız y tahmin değerleri arasındaki hata miktarı kabaca 1.64 olarak hesaplandı.*

In [25]:
#Test hatasının hesaplanması

In [26]:
rmse = np.sqrt(mean_squared_error(y_test, model.predict(X_test)))

In [27]:
rmse

np.float64(1.78159966153345)

*Eğitim seti ile eğittiğimiz modeli test setindeki hata hesaplama işleminde kullandık. Yani hiç görmediği gerçek y değerleri ile test setindeki bağımsız değişkenlerin ürettiği tahmini y değerleri arasındaki hata miktarı 1.78.*

 ## Model Tuning / Model Doğrulama

*Çoklu Doğrusal Regresyon'da dışsal olarak görülebilecek bir parametre olmadığı için burada Model Tuning'i aslında Model Doğrulama olarak ele alacağız.*

In [28]:
df.head()

,TV,radio,newspaper,sales
0,230.1,37.8,69.2,22.1
1,44.5,39.3,45.1,10.4
2,17.2,45.9,69.3,9.3
3,151.5,41.3,58.5,18.5
4,180.8,10.8,58.4,12.9


In [29]:
X = df.drop('sales', axis = 1)
y = df['sales']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 144)

lm = LinearRegression()
model = lm.fit(X, y)

In [30]:
# tahmin edilen eğitim hatası

In [31]:
np.sqrt(mean_squared_error(y_train, model.predict(X_train)))

np.float64(1.6785883805122659)

In [32]:
# tahmin edilen test hatası

In [33]:
np.sqrt(mean_squared_error(y_test, model.predict(X_test)))

np.float64(1.6278807596969487)

In [34]:
#doğrulanmış eğitim hatası

In [35]:
np.sqrt(-cross_val_score(model,
                X_train,
                y_train, 
                cv = 10,
                scoring = "neg_mean_squared_error")).mean()

np.float64(1.6649345607872932)

In [36]:
#doğrulanmış test hatası

In [37]:
np.sqrt(-cross_val_score(model,
                X_test,
                y_test, 
                cv = 10,
                scoring = "neg_mean_squared_error")).mean()

np.float64(1.7399924960346644)

*Burada dikkat etmemiz gereken birkaç husus var: Birincisi eğitim hatasının test hatasından farklılaşıyor olacağı ve eğitim hatasının test hatasının kötü bir tahmincisi olduğudur.*

*cross validation işlemi ile eğitim hatasını ve test hatasını doğrulamaya çalışıyoruz yani valide edilmiş hata değerleri daha güvenilirdir.*

*Bundan sonra model oluştururken eğitim setini valide ederek hatalara erişmek ve valide edilmiş modelin katsayılarını yani parametrelerini kullanarak yeni modeller üretmek olacak.*